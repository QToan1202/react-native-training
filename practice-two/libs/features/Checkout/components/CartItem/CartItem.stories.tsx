import { ComponentMeta, ComponentStory } from '@storybook/react-native'

import { CartItem } from '../index'

export default {
  title: 'components/Cart Item',
  component: CartItem,
} as ComponentMeta<typeof CartItem>

const Template: ComponentStory<typeof CartItem> = (args) => <CartItem {...args} />

export const Default = Template.bind({})
Default.args = {
  image: require('@assets/cart/item.png'),
  name: 'coca cola',
  price: 50,
  quantity: 1,
}

export const Item = Template.bind({})
Item.args = {
  image: require('@assets/store/tradly.png'),
  name: 'vegetables',
  price: 100,
  discountPrice: 2,
  quantity: 1,
}
