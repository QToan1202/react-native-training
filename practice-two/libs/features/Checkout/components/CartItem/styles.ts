import { GetProps, styled } from 'tamagui'
import { Image } from 'react-native'

import { Paragraph } from '@practice-two/shared'

export const StyledText = styled(Paragraph, {
  content: '',
  fontWeight: '$2',
  color: '$color.gray_50',
  textTransform: 'capitalize',
})

export type StyledTextProps = GetProps<typeof StyledText>

export const StyledImage = styled(Image, {
  source: { uri: '' },
  width: 102,
  height: 102,
  borderRadius: 10, // getTokenValue('$radius.5')
})

export type StyledImageProps = GetProps<typeof StyledImage>
