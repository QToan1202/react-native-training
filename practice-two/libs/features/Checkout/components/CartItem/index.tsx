import { memo } from 'react'
import isEqual from 'react-fast-compare'
import { ImageSourcePropType } from 'react-native'
import { XStack, XStackProps, YStack } from 'tamagui'

import { Paragraph, calculateDiscount } from '@practice-two/shared'

import { StyledImage, StyledText } from './styles'

export type CartItemProps = XStackProps & {
  image: ImageSourcePropType
  name: string
  price: number
  discountPrice?: number
  quantity: number
}

const CartItem = ({ image, name, price, discountPrice, quantity, ...rest }: CartItemProps) => {
  return (
    <XStack
      paddingTop={29}
      paddingLeft="$space.3.5"
      paddingBottom="$space.3"
      backgroundColor="$color.white"
      pressStyle={{ opacity: 0.5 }}
      {...rest}
    >
      <XStack space="$space.3.5">
        <StyledImage source={image} />
        <YStack space="$space.3" justifyContent="center">
          <StyledText content={name} />
          <XStack alignItems="center" columnGap="$space.1.5">
            <Paragraph
              content={`$${discountPrice ? discountPrice.toFixed(0) : price.toFixed(0)}`}
              fontWeight="$4"
              fontSize="$3"
              lineHeight="$4"
              color="$color.primary"
              numberOfLines={1}
            />
            {discountPrice !== 0 && (
              <>
                <StyledText
                  content={`$${price.toFixed(0)}`}
                  textDecorationLine="line-through"
                  numberOfLines={1}
                />
                <StyledText content={calculateDiscount(price, discountPrice)} numberOfLines={1} />
              </>
            )}
          </XStack>
          <StyledText content={`qty: ${quantity}`} />
        </YStack>
      </XStack>
    </XStack>
  )
}

export default memo(CartItem, isEqual)
