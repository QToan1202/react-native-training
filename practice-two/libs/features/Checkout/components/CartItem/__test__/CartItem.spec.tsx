import { render, screen } from '@testing-library/react-native'

import { COLORS } from '@constants'

import CartItem from '../index'

describe('Testing Cart item component', () => {
  beforeEach(() => {
    render(
      <CartItem
        image={require('@assets/cart/item.png')}
        name="coca cola"
        price={50}
        quantity={1}
        testID="cart-item"
      />
    )
  })

  it('Render Cart item successfully', () => {
    const component = render(
      <CartItem
        image={require('@assets/cart/item.png')}
        name="coca cola"
        price={50}
        discountPrice={25}
        quantity={1}
      />
    ).toJSON()

    expect(component).toMatchSnapshot()
  })

  it('Item must exist on the screen', () => {
    const component = screen.getByTestId('cart-item')

    expect(component).toBeOnTheScreen()
    expect(component).toBeVisible()
  })

  it('Item price must have the primary color', () => {
    const component = screen.getByText(/50/)

    expect(component).toHaveStyle({ color: COLORS.PRIMARY })
  })
})
