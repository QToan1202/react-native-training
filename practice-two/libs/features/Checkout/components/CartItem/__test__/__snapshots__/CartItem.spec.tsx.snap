// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Cart item component Render Cart item successfully 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#fff",
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "paddingTop": 29,
    }
  }
>
  <View
    style={
      {
        "flexDirection": "row",
        "gap": 15,
      }
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/cart/item.png",
        }
      }
      style={
        {
          "borderRadius": 10,
          "height": 102,
          "width": 102,
        }
      }
    />
    <View
      style={
        {
          "justifyContent": "center",
          "rowGap": 12.5,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#4f4f4f",
              "fontFamily": "Montserrat_500Medium",
              "textTransform": "capitalize",
            },
          ]
        }
      >
        coca cola
      </Text>
      <View
        style={
          [
            {
              "alignItems": "center",
              "columnGap": 4,
              "flexDirection": "row",
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              {
                "color": "#33907C",
                "fontFamily": "Montserrat_700Bold",
                "fontSize": 18,
              },
            ]
          }
        >
          $25
        </Text>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              [
                {
                  "color": "#4f4f4f",
                  "fontFamily": "Montserrat_500Medium",
                  "textTransform": "capitalize",
                },
                {
                  "textDecorationLine": "line-through",
                },
              ],
            ]
          }
        >
          $50
        </Text>
        <Text
          numberOfLines={1}
          style={
            [
              {
                "color": "#fff",
                "fontFamily": "Montserrat_400Regular",
              },
              {
                "fontSize": 14,
              },
              {
                "color": "#4f4f4f",
                "fontFamily": "Montserrat_500Medium",
                "textTransform": "capitalize",
              },
            ]
          }
        >
          50% off
        </Text>
      </View>
      <Text
        style={
          [
            {
              "color": "#fff",
              "fontFamily": "Montserrat_400Regular",
            },
            {
              "fontSize": 14,
            },
            {
              "color": "#4f4f4f",
              "fontFamily": "Montserrat_500Medium",
              "textTransform": "capitalize",
            },
          ]
        }
      >
        qty: 1
      </Text>
    </View>
  </View>
</View>
`;
